File: day2.rb
#!/usr/bin/env ruby

require "pry-byebug"

def input
  @input ||= DATA.read.split("\n")
end

class GameSubset
  attr_accessor(:red, :green, :blue)

  def initialize(red: 0, green: 0, blue: 0)
    @red = red
    @green = green
    @blue = blue
  end

  def valid?
    return false if red > 12
    return false if green > 13
    return false if blue > 14

    true
  end
end

class Game
  attr_reader(:id, :subsets)

  def initialize(id:, subsets:)
    @id = id
    @subsets = subsets
  end
end

def parse_games
  games = []

  input.each do |line|
    first_part, last_part = line.split(": ")

    game_id = first_part.match(/\d+/)[0].to_i

    game = Game.new(id: game_id, subsets: [])

    last_part.split("; ").each do |subset|
      matches = subset.scan(/(?<count>\d+)\s+(?<color>red|green|blue)/)

      game_subset = GameSubset.new

      matches.each do |(num, color)|
        game_subset.send("#{color}=", num.to_i)
      end

      game.subsets << game_subset
    end

    games << game
  end

  games
end

def part1
  games = parse_games

  games.sum do |game|
    if game.subsets.all?(&:valid?)
      game.id
    else
      0
    end
  end
end

puts "part1: #{part1}"

__END__
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green

File: day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type GameSubset struct {
	red   int
	green int
	blue  int
}

type Game struct {
	id      int
	subsets []GameSubset
}

// Lets use a more verbose bufio strategy instead of io/ioutil get newline
// avoidance for free.
func readInput(fileName string) ([]string, error) {
	file, err := os.Open(fileName)
	if err != nil {
		return nil, err
	}

	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return lines, nil
}

func part1(input []string) string {
	// Go through each line of input
	// For each line, parse out the game into information segmented by:
	// Game {
	//   id int
	//   []subset {
	//     red int
	//     green int
	//     blue int
	//   }
	// }

	// The Elf would first like to know which games would have been possible if
	// the bag contained only 12 red cubes, 13 green cubes, and 14 blue cubes?
	//
	// Go through each subset for each game, fail fast if the bag exceeds those ^
	// If they succeed, add gameId to counter
	// return sum

	games := []Game{}
	for _, line := range input {
		parts := strings.Split(line, ":")

		r := regexp.MustCompile(`\d+`)

		gameId, err := strconv.Atoi(r.FindString(parts[0]))

		if err != nil {
			panic(err)
		}

		subsets := []GameSubset{}

        for _, subsetPart := range strings.Split(parts[1], ";") {
            fmt.Println(subsetPart)
        }

        return ""


		// Imagine in a for loop

		subset := GameSubset{
			red:   1,
			blue:  2,
			green: 3,
		}
		subsets = append(subsets, subset)

		game := Game{
			id:      gameId,
			subsets: subsets,
		}

		games = append(games, game)
	}

	fmt.Println(games)

	return ""
}

func main() {
	input, err := readInput("./day1_input.txt")

	if err != nil {
		panic(err)
	}

	fmt.Println(fmt.Sprintf("Part 1: %s", part1(input)))
}
